### DISCUSSION TITLE
How do I use custom row actions with `editingMode='row'`?

### DISCUSSION THREAD

**Discussion by githorse** [KevinVandy/material-react-table#607] [Q&A]
The [custom row actions example](https://www.material-react-table.com/docs/guides/row-actions#row-actions-buttons-source-code) in the docs shows an example of an edit button that triggers the modal editor. This works because the modal editor itself knows how to save the data.

But I want to use custom row actions to trigger an [inline row edit](https://www.material-react-table.com/docs/guides/editing#row-editing-mode). To do that, I need to:

1. Inside `renderRowActions`, detect whether the row is currently being edited or not, so I can flip the edit button to a save button; and
2. In the save button, call some method to trigger the save (like the modal editor does).

Basically, I need two methods along the lines of `row.isEditing()` and `table.saveRow(row)`, I think:

```
renderRowActions=(({row, table}) => (
  row.isEditing() ? (
    <SaveButton onClick={() => table.saveRow(row)}
  ) : (
    <EditButton onClick={() => table.setEditingRow(row)}
  )
)}
```

(Maybe methods like these are available -- I looked but may have missed them.)

So how do I (1) detect that a row is in editing mode, and (2) trigger a save on that row? More generally, how do I use a custom edit button in inline editing mode?

**Comment by githorse** [comment]
Here is my latest attempt at this:

```
renderRowActions: ({row, table}) => {
  const originalRecord = row.original
  const editedRecord   = {...originalRecord, ...table.getState().editingRow?._valuesCache as Thingy}
  const startEditing   = () => table.setEditingRow(row)
  const stopEditing    = () => table.setEditingRow(null)
  const hasChanges     = !isEqual(originalRecord, editedRecord)
  const isEditing      = table.getState().editingRow?.id === row.id
  const index          = row.index
  return (
    isEditing ? (
      <Stack>
        <SaveButton onClick={() => saveThingy(editedRecord)} disabled={!hasChanges} />
        <CancelButton onClick={stopEditing} />
      </Stack>
    ) : (
      <Stack>
        <EditButton onClick={startEditing} />
        <DeleteButton onClick={() => deleteThingy(index)} />
      </Stack>
    )
  )
}
```

Obviously, this is pretty convoluted; perhaps future MRT versions might export higher level abstractions than `row` and `table` (like all my `const`s) to make this easier. However, this doesn't quite work yet. It works with built-in editors, but not when I use a custom editing widget, in which case it gets glitchy and fails to save correctly. Still trying to figure out why, but feel free to weigh in. I'll edit this if I figure it out.
